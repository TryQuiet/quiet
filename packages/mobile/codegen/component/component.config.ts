import chalk from 'chalk';
import {Answers} from 'inquirer';
import {composeValidators} from '../composeValidators.helper';
import {PlopGeneratorConfig} from '../PlopGeneratorConfig.type';
import {requireInput} from '../requireInput.helper';
import {requirePascalCase} from '../requirePascalCase.helper';
import {componentActions} from './component.actions';
import {componentConst} from './component.const';

export const componentConfig: PlopGeneratorConfig = {
  actions: answers => {
    if (!answers || !answers[componentConst.vars.shouldGenerateCode]) {
      return [];
    }

    return componentActions;
  },

  description: 'Generate a component.',
  prompts: [
    {
      message: 'Component name (pascal case string):',
      name: componentConst.vars.name,
      type: 'input',
      validate: composeValidators(
        requireInput('Component name cannot be empty!'),
        requirePascalCase('Component name must be in pascal case!'),
      ),
    },
    {
      choices: [
        {name: 'Yes', value: true},
        {name: 'No', value: false},
      ],
      message: (answers: Answers): string =>
        `Do you want to generate ${chalk.green(
          answers[componentConst.vars.name],
        )} component?`,
      name: componentConst.vars.shouldGenerateCode,
      type: 'list',
    },
  ],
};
