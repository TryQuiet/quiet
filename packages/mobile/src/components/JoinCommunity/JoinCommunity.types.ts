export interface JoinCommunityProps {
  joinCommunityAction: (address: string) => void
  redirectionAction: () => void
  networkCreated: boolean
  invitationCode?: string
  ready?: boolean
}
