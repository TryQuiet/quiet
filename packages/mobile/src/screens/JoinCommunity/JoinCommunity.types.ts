import { JoinCommunityRouteProp } from '../../route.params'

export interface JoinCommunityScreenProps {
    route: JoinCommunityRouteProp
}
