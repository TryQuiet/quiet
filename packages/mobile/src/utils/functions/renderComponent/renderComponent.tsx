import { render, RenderAPI } from '@testing-library/react-native'
import React, { FC, ReactElement } from 'react'
import { ThemeProvider } from 'styled-components/native'

import { defaultTheme } from '../../../styles/themes/default.theme'
import { RenderComponentProps } from './renderComponent.types'

export const renderComponent = (ui: ReactElement): RenderAPI => {
  const Wrapper: FC<RenderComponentProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
  )

  return render(ui, { wrapper: Wrapper })
}
