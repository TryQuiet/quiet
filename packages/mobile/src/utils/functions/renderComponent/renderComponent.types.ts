export interface RenderComponentProps {
  children?: React.ReactNode
}
