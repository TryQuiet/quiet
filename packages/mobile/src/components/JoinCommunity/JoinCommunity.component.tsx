import React, { FC, useEffect, useState, useRef } from 'react'
import { Keyboard, KeyboardAvoidingView, TextInput, View } from 'react-native'
import { defaultTheme } from '../../styles/themes/default.theme'
import { Button } from '../Button/Button.component'
import { Input } from '../Input/Input.component'
import { Typography } from '../Typography/Typography.component'
import { TextWithLink } from '../TextWithLink/TextWithLink.component'

import { JoinCommunityProps } from './JoinCommunity.types'
import { getInvitationCode } from '@quiet/state-manager'
import { ONION_ADDRESS_REGEX } from '@quiet/common'

export const JoinCommunity: FC<JoinCommunityProps> = ({ joinCommunityAction, redirectionAction, invitationCode, networkCreated }) => {
  const [joinCommunityInput, setJoinCommunityInput] = useState<string | undefined>()
  const [inputError, setInputError] = useState<string | undefined>()
  const [loading, setLoading] = useState<boolean>(false)

  const inputRef = useRef<TextInput>()

  const onChangeText = (value: string) => {
    setInputError(undefined)
    setJoinCommunityInput(value)
  }

  const onPress = () => {
    Keyboard.dismiss()
    setLoading(true)

    let submitValue: string = joinCommunityInput

    if (submitValue === undefined || submitValue?.length === 0) {
      setLoading(false)
      setInputError('Community address can not be empty')
      return
    }

    submitValue = getInvitationCode(submitValue.trim())
    if (!submitValue || !submitValue.match(ONION_ADDRESS_REGEX)) {
      setLoading(false)
      setInputError('Please check your invitation code and try again')
      return
    }

    joinCommunityAction(submitValue)
  }

  useEffect(() => {
    if (invitationCode) {
      setJoinCommunityInput(invitationCode)
      setInputError(undefined)
      setLoading(true)
      inputRef.current?.setNativeProps({ text: invitationCode })
    }
  }, [invitationCode])

  useEffect(() => {
    if (networkCreated) {
      setInputError(undefined)
      setJoinCommunityInput('')
    }
  }, [networkCreated])

  return (
    <View style={{ flex: 1, backgroundColor: defaultTheme.palette.background.white }}>
      <KeyboardAvoidingView
        behavior='height'
        style={{
          flex: 1,
          justifyContent: 'center',
          paddingLeft: 20,
          paddingRight: 20
        }}>
        <Typography fontSize={24} fontWeight={'medium'} style={{ marginBottom: 30 }}>
          {'Join community'}
        </Typography>
        <Input
          onChangeText={onChangeText}
          label={'Paste your invite link to join an existing community'}
          placeholder={'Invite link'}
          disabled={loading}
          validation={inputError}
          ref={inputRef}
        />
        <View style={{ marginTop: 32 }}>
          <TextWithLink
            text={'You can %a instead'}
            links={[
              {
                tag: 'a',
                label: 'create a new community',
                action: redirectionAction
              }
            ]}
          />
        </View>
        <View style={{ marginTop: 32 }}>
          <Button onPress={onPress} title={'Continue'} loading={loading} />
        </View>
      </KeyboardAvoidingView>
    </View>
  )
}
