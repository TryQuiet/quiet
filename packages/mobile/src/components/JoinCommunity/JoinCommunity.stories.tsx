import { storiesOf } from '@storybook/react-native'
import React from 'react'
import { storybookLog } from '../../utils/functions/storybookLog/storybookLog.function'
import { JoinCommunity } from './JoinCommunity.component'

storiesOf('JoinCommunity', module).add('Default', () => (
  <JoinCommunity
    joinCommunityAction={storybookLog('Opening username registration!')}
    redirectionAction={storybookLog('Navigating to create community')}
    networkCreated={false}
  />
))
