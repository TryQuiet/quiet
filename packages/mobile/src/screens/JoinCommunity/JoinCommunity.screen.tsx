/* eslint-disable padded-blocks */
import React, { FC, useCallback, useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { identity, communities } from '@quiet/state-manager'
import { CommunityOwnership, CreateNetworkPayload } from '@quiet/types'
import { JoinCommunity } from '../../components/JoinCommunity/JoinCommunity.component'
import { navigationActions } from '../../store/navigation/navigation.slice'
import { ScreenNames } from '../../const/ScreenNames.enum'
import { JoinCommunityScreenProps } from './JoinCommunity.types'

export const JoinCommunityScreen: FC<JoinCommunityScreenProps> = ({ route }) => {
  const dispatch = useDispatch()

  const [invitationCode, setInvitationCode] = useState<string | undefined>(undefined)

  const currentCommunity = useSelector(communities.selectors.currentCommunity)
  const currentIdentity = useSelector(identity.selectors.currentIdentity)
  const networkCreated = Boolean(currentCommunity && !currentIdentity?.userCertificate)

  const community = useSelector(communities.selectors.currentCommunity)

  // Handle deep linking (opening app with quiet://)
  useEffect(() => {
    const code = route.params?.code

    // Screen hasn't been open through a link
    if (!code) return

    // Change component state
    setInvitationCode(code)
  }, [dispatch, community, route.params?.code])

  useEffect(() => {
    if (networkCreated) {
      dispatch(
        navigationActions.navigation({
          screen: ScreenNames.UsernameRegistrationScreen,
        })
      )
    }
  }, [dispatch, currentCommunity])

  const joinCommunityAction = useCallback(
    (address: string) => {
      const payload: CreateNetworkPayload = {
        ownership: CommunityOwnership.User,
        registrar: address,
      }
      dispatch(communities.actions.createNetwork(payload))
    },
    [dispatch]
  )

  const redirectionAction = useCallback(() => {
    dispatch(
      navigationActions.navigation({
        screen: ScreenNames.CreateCommunityScreen,
      })
    )
  }, [dispatch])

  return (
    <JoinCommunity
      joinCommunityAction={joinCommunityAction}
      redirectionAction={redirectionAction}
      networkCreated={networkCreated}
      invitationCode={invitationCode}
    />
  )
}
