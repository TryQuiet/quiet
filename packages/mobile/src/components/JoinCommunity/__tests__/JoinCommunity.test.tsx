import { renderComponent } from '../../../utils/functions/renderComponent/renderComponent'
import { JoinCommunity } from '../JoinCommunity.component'

describe('JoinCommunity component', () => {
    it('renders component', () => {
        const { toJSON } = renderComponent(
            <JoinCommunity joinCommunityAction={jest.fn()} redirectionAction={jest.fn()} networkCreated={false} />
        )
        expect(toJSON()).toMatchSnapshot()
    })

    it('renders loading screen if not ready', () => {
        const { toJSON } = renderComponent(
            <JoinCommunity
                joinCommunityAction={jest.fn()}
                redirectionAction={jest.fn()}
                networkCreated={false}
                ready={false}
            />
        )
        expect(toJSON()).toMatchSnapshot()
    })
})
