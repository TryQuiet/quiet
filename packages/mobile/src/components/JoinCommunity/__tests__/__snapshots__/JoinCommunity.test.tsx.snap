// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinCommunity component renders component 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
  testID="join-community-component"
>
  <View
    onLayout={[Function]}
    style={
      {
        "flex": 1,
        "justifyContent": "center",
        "paddingLeft": 20,
        "paddingRight": 20,
      }
    }
  >
    <Text
      color="main"
      fontSize={24}
      fontWeight="medium"
      horizontalTextAlign="left"
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "Rubik-Medium",
            "fontSize": 24,
            "textAlign": "left",
            "textAlignVertical": "center",
          },
          {
            "marginBottom": 30,
          },
        ]
      }
      verticalTextAlign="center"
    >
      Join community
    </Text>
    <View>
      <Text
        color="main"
        fontSize={14}
        horizontalTextAlign="left"
        style={
          [
            {
              "color": "#000000",
              "fontFamily": "Rubik-Regular",
              "fontSize": 14,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            {
              "color": "#4C4C4C",
              "paddingBottom": 10,
            },
          ]
        }
        verticalTextAlign="center"
      >
        Paste your invite link to join an existing community
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        round={false}
        style={
          [
            {
              "backgroundColor": "#ffffff",
              "borderColor": "#C4C4C4",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexGrow": 1,
              "height": 56,
              "justifyContent": "center",
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            {
              "height": 54,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          height={54}
          onChangeText={[Function]}
          onContentSizeChange={[Function]}
          placeholder="Invite link"
          style={
            [
              {
                "height": 54,
                "paddingBottom": 12,
                "paddingTop": 12,
                "textAlignVertical": "center",
              },
            ]
          }
          testID="input"
        />
      </View>
    </View>
    <View
      style={
        {
          "marginTop": 32,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "display": "flex",
            "flex": 1,
            "flexDirection": "row",
            "flexWrap": "wrap",
            "minHeight": 500,
          }
        }
      >
        <Text
          color="main"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
          You
        </Text>
        <Text
          color="main"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
           
        </Text>
        <Text
          color="main"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
          can
        </Text>
        <Text
          color="main"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
           
        </Text>
        <Text
          color="main"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
          <Text
            color="link"
            fontSize={14}
            horizontalTextAlign="left"
            onPress={[Function]}
            style={
              [
                {
                  "color": "#67BFD3",
                  "fontFamily": "Rubik-Regular",
                  "fontSize": 14,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
              ]
            }
            verticalTextAlign="center"
          >
            create a new community
          </Text>
        </Text>
        <Text
          color="main"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
           
        </Text>
        <Text
          color="main"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
          instead
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "marginTop": 32,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#521C74",
            "borderRadius": 5,
            "justifyContent": "center",
            "marginVertical": 12,
            "minHeight": 45,
            "paddingVertical": 12,
            "width": undefined,
          }
        }
        testID="button"
      >
        <Text
          color="white"
          fontSize={14}
          horizontalTextAlign="left"
          style={
            [
              {
                "color": "#ffffff",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
            ]
          }
          verticalTextAlign="center"
        >
          Continue
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`JoinCommunity component renders loading screen if not ready 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="loading"
>
  <Image
    source={
      {
        "testUri": "../../../assets/icons/quiet_icon.png",
      }
    }
    style={
      {
        "borderRadius": 16,
        "height": 84,
        "marginBottom": 46,
        "marginTop": 20,
        "resizeMode": "cover",
        "width": 84,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "gap": 6,
      }
    }
  >
    <Text
      color="main"
      fontSize={14}
      fontWeight="medium"
      horizontalTextAlign="left"
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "Rubik-Medium",
            "fontSize": 14,
            "textAlign": "left",
            "textAlignVertical": "center",
          },
        ]
      }
      verticalTextAlign="center"
    >
      Starting backend
    </Text>
    <Text
      color="gray50"
      fontSize={12}
      horizontalTextAlign="left"
      style={
        [
          {
            "color": "#7F7F7F",
            "fontFamily": "Rubik-Regular",
            "fontSize": 12,
            "textAlign": "left",
            "textAlignVertical": "center",
          },
        ]
      }
      verticalTextAlign="center"
    >
      This can take some time
    </Text>
  </View>
  <View
    style={
      {
        "margin": 20,
      }
    }
  >
    <Text
      color="grayDark"
      fontSize={12}
      horizontalTextAlign="left"
      style={
        [
          {
            "color": "#999999",
            "fontFamily": "Rubik-Regular",
            "fontSize": 12,
            "textAlign": "left",
            "textAlignVertical": "center",
          },
        ]
      }
      verticalTextAlign="center"
    >
      v 1.0.0
    </Text>
  </View>
</View>
`;
